sdk/testing/stepwise/stepwise_test.go

```diff
@@ -0,0 +1,282 @@
+				}
+			}
+
+			if expectedT.ErrorCalled != testT.ErrorCalled {
+				t.Fatalf("expected ErrorCalled (%t), got (%t)", expectedT.ErrorCalled, testT.ErrorCalled)
+			}
+		})
+	}
+}
+
+func TestStepwise_makeRequest(t *testing.T) {
+	me := new(mockEnvironment)
+	me.Setup()
+	testT := new(mockT)
+
+	type testRequest struct {
+		Operation         StepOperation
+		Path              string
+		ExpectedRequestID string
+		ExpectErr         bool
+	}
+	testRequests := map[string]testRequest{
+		"list": testRequest{
+			Operation:         ListOperation,
+			Path:              "keys",
+			ExpectedRequestID: "list-request",
+		},
+		"read": testRequest{
+			Operation:         ReadOperation,
+			Path:              "keys/name",
+			ExpectedRequestID: "read-request",
+		},
+		"error": testRequest{
+			Operation: ReadOperation,
+			Path:      "error",
+			ExpectErr: true,
+		},
+	}
+
+	for name, tc := range testRequests {
+		t.Run(name, func(t *testing.T) {
+			step := Step{
+				Operation: tc.Operation,
+				Path:      tc.Path,
+			}
+
+			secret, err := makeRequest(testT, me, step)
+			if err != nil && !tc.ExpectErr {
+				t.Fatalf("unexpected error: %s", err)
+			}
+			if err == nil && tc.ExpectErr {
+				t.Fatal("expected error but got none:")
+			}
+
+			if err != nil && tc.ExpectErr {
+				return
+			}
+			if secret.RequestID != tc.ExpectedRequestID {
+				t.Fatalf("expected (%s), got (%s)", tc.ExpectedRequestID, secret.RequestID)
+			}
+		})
+	}
+}
+
+type mockEnvironment struct {
+	ts     *httptest.Server
+	client *api.Client
+	l      sync.Mutex
+
+	teardownCalled bool
+}
+
+// Setup creates the mock environment, establishing a test HTTP server
+func (m *mockEnvironment) Setup() error {
+	mux := http.NewServeMux()
+	// LIST
+	mux.HandleFunc("/v1/test/keys", func(w http.ResponseWriter, req *http.Request) {
+		listResp := api.Secret{
+			RequestID: "list-request",
+			LeaseID:   "lease-id",
+		}
+		out, err := jsonutil.EncodeJSON(listResp)
+		if err != nil {
+			panic(err)
+		}
+		w.Write(out)
+	})
+	// lease revoke
+	mux.HandleFunc("/v1/sys/leases/revoke", func(w http.ResponseWriter, req *http.Request) {
+		w.Write(nil)
+	})
+	// READ
+	mux.HandleFunc("/v1/test/keys/name", func(w http.ResponseWriter, req *http.Request) {
+		readResp := api.Secret{
+			RequestID: "read-request",
+			LeaseID:   "lease-id",
+		}
+		out, err := jsonutil.EncodeJSON(readResp)
+		if err != nil {
+			panic(err)
+		}
+		w.Write(out)
+	})
+	// fall through that rejects any other url than "/"
+	mux.HandleFunc("/", func(w http.ResponseWriter, req *http.Request) {
+		if req.URL.Path != "/" {
+			http.NotFound(w, req)
+			return
+		}
+		fmt.Fprintf(w, "{}")
+	})
+	m.ts = httptest.NewServer(mux)
+
+	return nil
+}
+
+// Client creates a Vault API client configured to the mock environment's test
+// server
+func (m *mockEnvironment) Client() (*api.Client, error) {
+	m.l.Lock()
+	defer m.l.Unlock()
+	// this shouldn't be needed but being defensive
```

Comment:
Should this be an assertion? Would it be valuable to know that this is being called?

Reply:
No, I don't think it should or that would be valuable. I changed it in fb96110b7 to simply return `nil` with an error if `ts` is `nil`.